src/main/java/com/bootcamp/demo/service/BookingService.java

```diff
@@ -0,0 +1,112 @@
+package com.bootcamp.demo.service;
+
+import com.bootcamp.demo.model.Booking;
+import com.bootcamp.demo.model.PaymentStatus;
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import org.springframework.stereotype.Service;
+
+import java.util.Comparator;
+import java.util.LinkedHashSet;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+@Service
+public class BookingService implements IBookingService {
+    private static final Logger LOGGER = Logger.getLogger(BookingService.class.getName());
+    private static final String COLLECTION_PATH = "bookings/databases/bookings";
+
+    private final Firestore db;
+
+    public BookingService(Firestore db) {
+        this.db = db;
+    }
+
+    /**
+     * Retrieves a Booking entity by its id.
+     */
+    @Override
+    public Booking getBookingByID(final String bookingId) throws ExecutionException, InterruptedException, IllegalArgumentException {
+        if (bookingId == null) throw new IllegalArgumentException("Parameter 'bookingId' cannot be null");
+        DocumentReference documentReference = db.collection(COLLECTION_PATH)
+                .document(bookingId);
+        ApiFuture<DocumentSnapshot> future = documentReference.get();
+        DocumentSnapshot document = future.get();
+        Booking booking;
+        if (document.exists()) {
+            booking = document.toObject(Booking.class);
+            return booking;
+        }
+        return null;

```

Comment:
You could simply chain the methods directly, without using intermediate variables (since if the doc does not exists, calling `toObject` on it will return null). Thus, you could simplify to:
```
return db.collection(COLLECTION_PATH)
        .document(bookingId)
        .get()
        .get()
        .toObject(Booking.class);
```